File: src/main/java/com/alexperal/maven/MavenParser.java
package com.alexperal.maven;

import java.util.concurrent.BlockingQueue;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MavenParser {

    private final static Predicate<String> START_MATCHER = Pattern.compile("maven-dependency-plugin:.*:tree .* @ maventool ---").asPredicate();
    private final static Predicate<String> END_MATCHER = Pattern.compile("------------------------------------------------------------------------").asPredicate();
    private final static Pattern DEPENDENCY_PATTERN = Pattern.compile("\\+\\- (.*):(.*):(.*):(.*)$");
    public static String poisonPill = new String();

    public MavenProject parse(BlockingQueue<String> queue) {
        MavenProject mavenProject = new MavenProject();
        String line;
        try {
            boolean started = false;
            while ((line = queue.take()) != poisonPill) {
                if (!started) {
                    started = checkIfLineIsStarted(line);
                } else {
                    if (isEnd(line)) {
                        break;
                    }
                    handleDependency(line, mavenProject);
                }
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return mavenProject;
    }

    private void handleDependency(String line, MavenProject mavenProject) {
        String dependency = line.replaceFirst("\\[\\w*\\] ", "");
        Matcher matcher = DEPENDENCY_PATTERN.matcher(dependency);
        if (matcher.matches()) {
            String group = matcher.group(1);
            String artifact = matcher.group(2);
            String type = matcher.group(3);
            String version = matcher.group(4);
            mavenProject.addRootDependency(new Dependency(group,artifact,version,type));
        } else {
            System.out.println(dependency);
        }
    }

    private boolean isEnd(String line) {
        return END_MATCHER.test(line);
    }

    private boolean checkIfLineIsStarted(String line) {
        return START_MATCHER.test(line);
    }
}

File: src/main/java/com/alexperal/maven/MavenParser.kt
package com.alexperal.maven

import java.util.*
import java.util.concurrent.BlockingQueue
import java.util.regex.Pattern

class MavenParser {

    fun parse(queue: BlockingQueue<String>): MavenProject {
        val mavenProject = MavenProject()
        val dependencyStack = Stack<Dependency>()
        var started = false
        var line: String = queue.take()
        while (line !== poisonPill) {
            if (!started) {
                started = checkIfLineIsStarted(line)
            } else {
                if (isEnd(line)) {
                    break
                }
                handleDependency(line, mavenProject, dependencyStack)
            }
            line = queue.take();
        }
        return mavenProject
    }

    private fun handleDependency(line: String, mavenProject: MavenProject, dependencyStack: Stack<Dependency>) {
        val dependency = replaceInfoLine(line)
        when {
            isProject(dependency) -> processProject(dependency, mavenProject)
            isDependency(dependency) -> processRootDependency(dependency, mavenProject, dependencyStack)
        }
        println(dependency)
    }

    private fun processRootDependency(line: String, mavenProject: MavenProject, dependencyStack: Stack<Dependency>) {
        val matcher = DEPENDENCY_PATTERN.matcher(line)
        if (matcher.matches()) {
            val depthLevel = dependencyDepthLevel(matcher.group(1), line)
            var lastDependencyDepth = dependencyStack.size
            val dependency = getDependency(line)
            while (depthLevel < lastDependencyDepth) {
                dependencyStack.pop()
                lastDependencyDepth = dependencyStack.size
            }
            val fatherDependency = if (depthLevel == 0) null else dependencyStack.peek()
            mavenProject.addDependency(fatherDependency, dependency)
            dependencyStack.push(dependency)
            if (depthLevel > lastDependencyDepth) {
                throw IllegalArgumentException()
            }
        }
    }

    private fun getDependency(dependency: String): Dependency {
        val conflictMatcher = OMITTED_DEPENDENCY_PATTERN.matcher(dependency);
        val matcher = if (conflictMatcher.matches()) {
            conflictMatcher
        } else {
            val matcher = DEPENDENCY_PATTERN.matcher(dependency);
            matcher.matches()
            matcher
        }
        val group = matcher.group(2)
        val artifact = matcher.group(3)
        val type = matcher.group(4)
        val version = matcher.group(5)
        val scope = matcher.group(6)
        return Dependency(group, artifact, version, type, scope)

    }

    private fun dependencyDepthLevel(depthLevel: String, line: String) = if (line.startsWith(" ")) ((depthLevel.count { it == ' ' } - 1) / 2) else (depthLevel.count { it == ' ' } / 2)

    private fun isDependency(dependency: String): Boolean {
        return !isProject(dependency)
    }

    private fun processProject(dependency: String, mavenProject: MavenProject) {
        val matcher = PROJECT_ID_PATTERN.matcher(dependency)
        if (matcher.matches()) {
            val group = matcher.group(1)
            val artifact = matcher.group(2)
            val type = matcher.group(3)
            val version = matcher.group(4)
            mavenProject.id = MavenProjectId(group, artifact, version, type)
        }
    }

    private fun isProject(dependency: String): Boolean {
        return !(dependency.startsWith('+') || dependency.startsWith('\\') || dependency.startsWith(' ') || dependency.startsWith('|'))
    }

    private fun replaceInfoLine(line: String) = line.replaceFirst("\\[\\w*\\] ".toRegex(), "")

    private fun isEnd(line: String): Boolean {
        return END_MATCHER.test(line)
    }

    private fun checkIfLineIsStarted(line: String): Boolean {
        return START_MATCHER.test(line)
    }

    companion object {

        private val START_MATCHER = Pattern.compile("maven-dependency-plugin:.*:tree .* @ maventool ---").asPredicate()
        private val END_MATCHER = Pattern.compile("------------------------------------------------------------------------").asPredicate()
        private val PROJECT_ID_PATTERN = Pattern.compile("(.*):(.*):(.*):(.*)$")
        private val DEPENDENCY_PATTERN = Pattern.compile("(.*) (.*):(.*):(.*):(.*):(.*)$")
        private val OMITTED_DEPENDENCY_PATTERN = Pattern.compile("(.*) \\((.*):(.*):(.*):(.*):(.*) - (.*)\\)$")
        var poisonPill = String()
    }
}
